mgr/src/server/plugins/volume_heal.cr

```diff
@@ -0,0 +1,380 @@
+      brick_data[brick.name] = brick
+    end
+    puts "br: #{brick_data.to_json}"
+  end
+
+  volume.distribute_groups.each do |dist_grp|
+    dist_grp.storage_units.each do |storage_unit|
+      puts "storage_unit name: #{storage_unit.node.name}, #{storage_unit.path}"
+
+      storage_unit_full_path = storage_unit.node.name + ":" + storage_unit.path
+
+      if brick_data.has_key?(storage_unit_full_path)
+        heal_storage_unit_data = brick_data[storage_unit_full_path]
+        storage_unit.metrics.heal_status = heal_storage_unit_data.status
+        storage_unit.metrics.heal_totalNumberOfEntries = heal_storage_unit_data.totalNumberOfEntries.to_i64
+        storage_unit.metrics.heal_numberOfEntriesInHealPending = 0
+        storage_unit.metrics.heal_numberOfEntriesInSplitBrain = 0
+        storage_unit.metrics.heal_numberOfEntriesPossiblyHealing = 0
+        # storage_unit.metrics.heal_pending_files = heal_storage_unit_data.heal_pending_files
+      end
+
+      puts "metrics: #{storage_unit.metrics.to_json}"
+    end
+  end
+  # output = output.strip.split("\n")
+
+  # puts "#{output}"
+
+  # storage_units = to_hash(output)
+
+  # out.each do |part|
+  #   next if part == ""
+  #   storage_unit = Hash(String, String).new
+  #   if part.split(" ")[0] == "Brick"
+  #     storage_unit[part.split(" ")[0]] = part.split(" ")[1]
+  #     storage_units << storage_unit
+  #     next
+  #   end
+  #   k, v = part.split(":")
+  #   storage_unit[k] = v
+  #   puts "storage_units at part #{storage_unit}, #{part}"
+  # end
+  # puts "storage units typeof: #{typeof(storage_units)}::: #{typeof(storage_units.to_json)}"
+
+  env.response.status_code = 201
+  # storage_units.to_json
+  volume.to_json
+end
+
+post "/api/v1/pools/:pool_name/volumes/:volume_name/heal_info" do |env|
+  pool_name = env.params.url["pool_name"]
+  volume_name = env.params.url["volume_name"]
+
+  next forbidden(env) unless Datastore.maintainer?(env.user_id, pool_name)
+
+  pool = Datastore.get_pool(pool_name)
+  if pool.nil?
+    halt(env, status_code: 400, response: ({"error": "The Pool(#{pool_name}) doesn't exists"}.to_json))
+  end
+
+  volume = Datastore.get_volume(pool_name, volume_name)
+  if volume.nil?
+    halt(env, status_code: 400, response: ({"error": "The Volume(#{volume_name}) doesn't exists"}.to_json))
+  end
+
+  if !volume.replicate_family?
+    halt(env, status_code: 400, response: ({"error": "Cannot heal Volume(#{volume_name}). It is non replicated"}.to_json))
+  end
+
+  # The volume must be mounted else, heal is not possible.
+  client_volfile = "/var/lib/kadalu/volfiles/client-dev-#{volume_name}.vol"
+
+  if !File.exists?(client_volfile)
+    halt(env, status_code: 400, response: ({"error": "The client volfile does not exist for volume(#{volume_name}). Mount and retry!"}.to_json))
+  end
+
+  puts "IN HEAL INFO"
+
+  rc, output, err = execute("/glusterfs/heal/src/glfsheal", ["vol1", "info-summary", "--xml", "volfile-path", client_volfile])
```

Comment:
Move execution and parsing to new function

Reply:
done